### PR TITLE
[minor] Support using template for Team, Config

### PR DESCRIPTION
**Pull request by sunny299**
**What's the change or impact of this PR?**:

- can apply team/config values from template as own values
- can override values by values in spec
- change objects  to read information from status.used instead of spec

**Related issue link**:

### CODE REVIEW

File: internal/samsahai/controller_test.go [hunk 1]
```diff
@@ -36,6 +50,16 @@ var _ = Describe("S2H Controller", func() {
 			g.Expect(name).To(Equal(pluginName))
 		})
 
+		It("should apply template to team correctly", func() {
```
Comment (quraynai) [inline internal/samsahai/controller_test.go:53]: Please add test for failure scenarios..

File: internal/config/controller_test.go [hunk 2]
```diff
@@ -105,6 +111,19 @@ var _ = Describe("Config Controller", func() {
 		}))
 	})
 
+	It("should apply template to config correctly", func() {
```
Comment (quraynai) [inline internal/config/controller_test.go:114]: Please add test for failure scenario

File: api/v1beta1/config_types.go [hunk 3]
```diff
@@ -399,12 +405,44 @@ type ConfigSpec struct {
...
 
 // ConfigStatus defines the observed state of Config
 type ConfigStatus struct {
+	// UsedTemplate contains override configuration spec
+	// +optional
+	Used ConfigSpec `json:"used,omitempty"`
+
+	// TemplateUID
+	// +optional
+	TemplateUID string `json:"templateUpdatedID,omitempty"`
```
[the remark this answers is not in the document]
Comment (Pohfy123) [reply api/v1beta1/config_types.go:424]: please add a definition as well

File: api/v1beta1/team_types.go [hunk 4]
```diff
@@ -110,6 +110,15 @@ type TeamStatus struct {
 	// ActivePromotedBy represents a person who promoted the ActivePromotion
 	// +optional
 	ActivePromotedBy string `json:"activePromotedBy,omitempty"`
+
+	// +optional
+	Used TeamSpec `json:"used,omitempty"`
+
+	// +optional
+	TemplateUID string `json:"templateUpdatedID,omitempty"`
```
Comment (Pohfy123) [inline api/v1beta1/team_types.go:118]: should use the same name as object? templateUID

Comment (Pohfy123) [reply api/v1beta1/team_types.go:118]: please add a definition as well

File: api/v1beta1/team_types.go [hunk 5]
```diff
@@ -251,6 +260,8 @@ const (
 	TeamPostPreActiveNamespaceCreationRun TeamConditionType = "TeamPostPreActiveNamespaceCreationRun"
 	TeamFirstNotifyComponentChanged       TeamConditionType = "TeamFirstNotifyComponentChanged"
 	TeamFirstActivePromotionRun           TeamConditionType = "TeamFirstActivePromotionRun"
+	TeamUsedUpdated                       TeamConditionType = "TeamUsedUpdated"
+	TeamValidatedRequireField             TeamConditionType = "TeamValidatedRequireField"
```
Comment (Pohfy123) [inline api/v1beta1/team_types.go:264]: Should be `TeamRequiredFieldsValidated`?

File: test/e2e/samsahai/ctrl.go [hunk 6]
```diff
@@ -1876,6 +1876,57 @@ var _ = Describe("[e2e] Main controller", func() {
...
+			if err = samsahaiCtrl.GetTeam(teamName, &team); err != nil {
+				return false, nil
+			}
+			team.Spec.StagingCtrl.Endpoint = "http://127.0.0.1"
+			if err = samsahaiCtrl.UpdateTeam(&team); err != nil {
+				return false, nil
+			}
+			if err = samsahaiCtrl.EnsureTeamTemplateChanged(&teamUsingTemplate); err != nil {
+				return false, nil
+			}
```
Comment (Pohfy123) [inline test/e2e/samsahai/ctrl.go:1919]: should verify only result not trigger internal function

File: test/e2e/samsahai/ctrl.go [hunk 7]
```diff
@@ -1876,6 +1876,57 @@ var _ = Describe("[e2e] Main controller", func() {
...
+		By("Creating Team")
+		team := mockTeam
+		Expect(client.Create(ctx, &team)).To(BeNil())
+
+		By("Creating Team using template")
+		teamUsingTemplate := mockTeam2
+		Expect(client.Create(ctx, &teamUsingTemplate)).To(BeNil())
+
+		By("Apply team template")
+		Expect(samsahaiCtrl.EnsureTeamTemplateChanged(&teamUsingTemplate))
```
Comment (Pohfy123) [inline test/e2e/samsahai/ctrl.go:1902]: should verify only result not trigger internal function

**Review by Pohfy123 — APPROVED** [review]

**Pull request merged**